infra/blueprint-test/pkg/bpt/bpt.go

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package bpt defines a blueprint and implements default stages and execution order for a blueprint test
+
+package bpt
+
+import (
+	"github.com/GoogleCloudPlatform/cloud-foundation-toolkit/infra/blueprint-test/pkg/utils"
+	"github.com/mitchellh/go-testing-interface"
+	"github.com/stretchr/testify/assert"
+)
+
+// Blueprint represents a config that can be initialized, applied, verified and torndown.
+type Blueprint interface {
+	Init(*assert.Assertions)
+	Apply(*assert.Assertions)
+	Verify(*assert.Assertions)
+	Teardown(*assert.Assertions)
+}
+
+// BlueprintTest implements a generic blueprint
+type BlueprintTest struct {
+	init     func(*assert.Assertions) // custom init function
+	apply    func(*assert.Assertions) // custom apply function
+	verify   func(*assert.Assertions) // custom verify function
+	teardown func(*assert.Assertions) // custom teardown function
+	bp       Blueprint                // blueprint to be tested
+}
+
+type bptOption func(*BlueprintTest)
+
+func DefineInit(init func(*assert.Assertions)) bptOption {
+	return func(bt *BlueprintTest) {
+		bt.init = init
+	}
+}
+
+func DefineApply(apply func(*assert.Assertions)) bptOption {
+	return func(bt *BlueprintTest) {
+		bt.apply = apply
+	}
+}
+
+func DefineTeardown(teardown func(*assert.Assertions)) bptOption {
+	return func(bt *BlueprintTest) {
+		bt.teardown = teardown
+	}
+}
+
+func DefineVerify(verify func(*assert.Assertions)) bptOption {
+	return func(bt *BlueprintTest) {
+		bt.verify = verify
+	}
+}
+
+func (bt *BlueprintTest) Init(a *assert.Assertions) {
+	if bt.init != nil {
+		bt.init(a)
+	} else {
+		bt.bp.Init(a)
+	}
+}
+
+func (bt *BlueprintTest) Apply(a *assert.Assertions) {
+	if bt.apply != nil {
+		bt.apply(a)
+	} else {
+		bt.bp.Apply(a)
+	}
+}
+
+func (bt *BlueprintTest) Teardown(a *assert.Assertions) {
+	if bt.teardown != nil {
+		bt.teardown(a)
+	} else {
+		bt.bp.Teardown(a)
+	}
+}
+
+func (bt *BlueprintTest) Verify(a *assert.Assertions) {
+	if bt.verify != nil {
+		bt.verify(a)
+	} else {
+		bt.bp.Verify(a)
+	}
+}
+
+// TestBlueprint runs init, apply, verify, teardown in order for a given blueprint
+func TestBlueprint(t testing.TB, bp Blueprint, opts ...bptOption) {
```

Comment:
I'm not sure we should use the options pattern for defining the stage overrides. Purely on a stylistic note, it's going to lead some very heavily nested code where the TestBlueprint() call has lots of options trailing off it.

I'm not sure there's much benefit to this over just making `DefineVerify` methods on the BlueprintTest.

Reply:
I felt like maintaining parity with other packages would make it easier for users. After testing a few examples, this option also seemed more understandable than interacting with BlueprintTest.

Previous:
```
bpt.TestBlueprint(t, nt, func(tft *bpt.BlueprintTest) {
		tft.DefineVerify(func(assert *assert.Assertions) {
...
		})
	})
```

Current:
```
bpt.TestBlueprint(t, nt,
		bpt.DefineVerify(func(assert *assert.Assertions) {
...
		}),
	)
```
 For most use cases we also only expect an override on verify. Happy to switch if you think otherwise

Reply:
@morgante refactored this as we discussed offline. I also renamed `tft.Init` to `tft.NewTFBlueprintTest` to avoid confusion with `TFBlueprintTest.Init`. Currently I have kept `TestBlueprint` in bpt as I thought we can reuse this for KRM as well.

Current example is like
```go
networkBlueprint := tft.NewTFBlueprintTest(t)
networkBlueprint.DefineVerify(
		func(assert *assert.Assertions) {
			networkBlueprint.DefaultVerify(assert)
			...other custom asserts
		})
bpt.TestBlueprint(t, networkBlueprint)
```

Reply:
You can reuse a method across both TFBlueprintTest and KRMBlueprintTest via an embedded struct. I suggest moving this to TFBlueprintTest for now so we can have this Ux:

```
networkBlueprint := tft.NewTFBlueprintTest(t)
networkBlueprint.DefineVerify(
		func(assert *assert.Assertions) {
			networkBlueprint.DefaultVerify(assert)
			...other custom asserts
		})
networkBlueprint.Test()
```

Reply:
@morgante done, will do a follow up for switching to an embedded struct.